Intervention approvals

mlcocdav opened the pull request:
## This PR contains:
- [x] New features
- [ ] Changes to dev-tools e.g. CI config / github tooling
- [x] Docs
- [ ] Bug fixes
- [ ] Code refactor

### What is the current behavior? (You can also link to an open issue here)
https://github.com/UKGovernmentBEIS/inspect_ai/issues/521

The current intervention mode example lacks sophisticated approval mechanisms for tool calls and agent actions.

### What is the new behavior?
This PR introduces sophisticated approval mechanisms to the existing intervention mode example:

1. Implements an AllowListApprover for automatic approval of predefined bash commands.
2. Adds a HumanApprover for manual approval, rejection, escalation, or termination of tool calls.
3. Creates an ApprovalManager to manage chains of approvers.
4. Modifies the bash tool to integrate with the approval system.
5. Updates the agent loop to support agent-level approval for all tool calls.
6. Adds two new tasks: tool_call_intervention and agent_loop_intervention.
7. Updates the README with a detailed explanation of new features and usage instructions.

These changes demonstrate how to integrate flexible and secure approval systems into Inspect agents, allowing for both automatic and human-in-the-loop control over agent actions.

### Does this PR introduce a breaking change? (What changes might users need to make in their application due to this PR?)
No, this PR does not introduce breaking changes. It adds new functionality without modifying existing behavior. Users can opt-in to use the new approval mechanisms in their applications.

### Other information:
This PR enhances the security and flexibility of the Inspect AI framework by providing more granular control over agent actions. It allows for both automatic approval of safe commands and human intervention for potentially risky operations, making it suitable for a wider range of use cases and security requirements.

jjallaire-aisi commented on examples/intervention_approvals/intervention.py:
```diff
@@ -0,0 +1,290 @@
+from rich.console import Group, RenderableType, Console
+
+from inspect_ai.model._chat_message import ChatMessageAssistant
+console = Console()
```
Rich has a global default console which you can access with `rich.get_console()`. That's the one Inspect uses so that's the one you should use here.

jjallaire-aisi commented on examples/intervention_approvals/approvals.py:
```diff
@@ -0,0 +1,335 @@
+import sys
+import shlex
+
+class ApprovalDecision(Enum):
```
We have been using `Literal["approve", "reject", "escalate", "terminate"]` for enums

jjallaire-aisi commented on examples/intervention_approvals/approvals.py:
```diff
@@ -0,0 +1,335 @@
+    ESCALATE = "escalate"
+    TERMINATE = "terminate"
+
+class Approver(ABC):
```
For "single method" classes like these we use `Protocol` (with a `__call__` method) throughout Inspect, as a `Protocol` can be implemented either as a class or a free function. See `Solver`, `Scorer`, etc. One important reason we do this is that everything in Inspect should be stateless (as all of the instances of classes get used many times over for different samples, etc.) and a free function is much more stateless in implementation by default (essentially just trying to take away a foot gun).

jjallaire-aisi commented on examples/intervention_approvals/approvals.py:
```diff
@@ -0,0 +1,335 @@
+            state (Optional[TaskState]): The current task state, if available.
+
+        Returns:
+            Tuple[ApprovalDecision, str]: A tuple containing the approval decision and a message explaining the decision.
```
I think here's I'd just have a BaseModel named Approval:

```python
class Approval(BaseModel):
     decision: Literal["approve", "reject", "escalate", "terminate"]
     explanation: str
```

jjallaire-aisi commented on examples/intervention_approvals/approvals.py:
```diff
@@ -0,0 +1,335 @@
+            return ApprovalDecision.TERMINATE, "Terminated by human approver."
+        return ApprovalDecision.ESCALATE, "Invalid input from human approver, escalating."
+
+class ApprovalManager:
```
Here again we really try to stay away from classes wherever possible. In this case the only instance method that accesses `self.approvers` is `get_approval`. You should re-write it as a free function:

```python
def get_approval(approvers: list[Approver], tool_call: ToolCall) -> Tuple[bool, str]: ...
```

Then also make the rest of the member free functions.

This might seem like a semantic nit, but we will be so very much happier if we don't accumulate state in an `ApprovalsManager`. If you think it is inevitable that we might in the future, I'd still say let's wait and see and cross that bridge when we come to it (we can always have an "approvals state" type that we pass to the functions).

jjallaire-aisi reviewed (commented):
Looks great! I have some stylistic comments about aligning better w/ the Inspect codebase but the functionality and core ideas look dead on.

mlcocdav commented:
Hi @jjallaire-aisi, thank you very much for your comments! This PR is now ready for review. All your previous comments should be addressed (e.g. using stateless patterns as much as possible) and I've added a bit more functionality and documentation. Happy to hear your comments. Here is a summary of key updates:

1. **Refactored Approvals System:**
   - Implemented Approver as a Protocol for increased flexibility
   - Created standardized Approval BaseModel
   - Converted class-based approvers to stateless functions

2. **Improved Agent Loop and Tool Call Execution:**
   - Modified agent_loop to process tool calls individually
   - Implemented per-tool-call approval checks
   - Added execution of approved tool calls one at a time
   - Introduced rejection and skipping of unapproved tool calls

3. **Enhanced Python Code Safety:**
   - Added `python_allowlist_approver` for stricter Python code checks
   - Implemented module and function allowlisting
   - Added protection against sensitive modules and disallowed built-ins

4. **Approver Tool Wrapping:**
   - Introduced `wrap_approvers()` for multi-stage approval process -> this allows wrapping tool functions without the need to change them
   - Applied approvers to bash and python functions from Inspect
   - Moved approval-related printing functions to utils.py

5. **Updated Documentation:**
   - Expanded README with detailed explanations of approver types and functionality
   - Added examples and screenshots of tool call interventions
   - Included code snippets for implementation and customization

jjallaire-aisi commented:
Thanks @mlcocdav! This looks terrific. I am just about to start working on integrating the approver concept into Inspect (was hoping to start Monday but got delayed by some other stuff). Here is roughly my plan:

1) Take the core Approval / Approver constructs more or less as-is and bring them into the `inspect.tool` module.

2) Add some sort of "context" metadata to `@tool` so tools can tag themselves as e.g. executing bash or python commands (or generic "code") or tag themselves as having open ended access to the web (e.g. the new web browser tool) or calling a REST API. e.g. something like this: `context: Literal["code", "bash", "python", "web", "api"]`. The idea is that some approval managers might want to only handle certain contexts (or only handle tool calls that have declared a context). This also provides a hint for specialised approvers like the bash and python ones you created.

3) Add the approval manager and human approver to core Inspect.

4) Provide a way to bind to approvers when doing `inspect eval`, `eval()`, etc. 

Then, I would modify the example to use the core Inspect constructs, and leave the bash and python approvers as examples (we may want to bake something like this in eventually but I'd like to see these develop a bit first based on field experience with them).

jjallaire-aisi commented:
@mlcocdav We now have a PR for an Inspect native version of this here: https://github.com/UKGovernmentBEIS/inspect_ai/pull/654

Will close this PR in lieu of that one (the bash and python allowlist approvers are both in a new `approvers/` example and the interaction mode example has been simplified substantially by using the standard `human` approver and the new trace mode).

mlcocdav commented:
@jjallaire-aisi That makes sense, thanks!
